File: server/player_code_runner.py
import sys
import json
import importlib

def main():
    if len(sys.argv) < 2:
        print("Usage: python player_code_runner.py <player_number>", file=sys.stderr)
        sys.exit(1)

    player_number = sys.argv[1]
    player_module_name = f"player{player_number}_py"

    try:
        # Import the parent player module
        player_module = importlib.import_module(player_module_name)
    except ModuleNotFoundError:
        print(f"Player module {player_module_name} not found.", file=sys.stderr)
        sys.exit(1)

    memory = {}

    # List of unit names to dynamically import from the player module
    unit_names = [
        "acid_ant", "bloated_bedbug", "dung_beetle", "engorged_tick", 
        "famished_tick", "foraging_maggot", "infected_mouse", "lava_ant", 
        "mantis", "mawing_beaver", "plague_bat", "rhino_beetle", 
        "spider", "swooping_bat", "tainted_cockroach", "tunneling_mole"
    ]

    # Dynamically import each unit module from the player module
    unit_functions = {}
    for unit_name in unit_names:
        try:
            unit_module = importlib.import_module(f"{player_module_name}.{unit_name}")
            # Assuming each module has a function with the same name as the module
            unit_functions[unit_name] = getattr(unit_module, unit_name)
        except (ModuleNotFoundError, AttributeError):
            print(f"Failed to import {unit_name} from {player_module_name}.", file=sys.stderr)

    while True:
        unit_state = json.loads(sys.stdin.readline().strip())
        unit_id = unit_state.get("self").get("id")
        unit_type = unit_state.get("self").get("type")

        if unit_id not in memory:
            memory[unit_id] = {}

        unit_function = unit_functions.get(unit_type)

        if unit_function:
            instruction = unit_function(unit_state, memory[unit_id])
            print(f"{unit_id}:{instruction}")
            sys.stdout.flush()
        else:
            print(f"Unknown unit type: {unit_type}", file=sys.stderr)

if __name__ == "__main__":
    main()

File: server/player_code_runner.cpp
#include <cassert>
#include <iostream>
#include <string>
#include <unordered_map>
#include <functional>

#include "json.hpp"

#include "player1/acid_ant.h"
#include "player1/bloated_bedbug.h"
#include "player1/dung_beetle.h"
#include "player1/engorged_tick.h"
#include "player1/famished_tick.h"
#include "player1/foraging_maggot.h"
#include "player1/infected_mouse.h"
#include "player1/lava_ant.h"
#include "player1/mantis.h"
#include "player1/mawing_beaver.h"
#include "player1/plague_bat.h"
#include "player1/rhino_beetle.h"
#include "player1/swooping_bat.h"
#include "player1/tainted_cockroach.h"
#include "player1/tunneling_mole.h"

using UnitFunction = std::function<std::tuple<int, int, float, std::string>(const nlohmann::json&, std::unordered_map<std::string, std::string>&)>;

int main(int argc, char** argv) {
    std::string unit_state_str;
    int unit_id;
    std::string player_number = argv[1];
    nlohmann::json unit_state;

    std::unordered_map<int, std::unordered_map<std::string, std::string>> memory;

    std::unordered_map<std::string, UnitFunction> unit_functions = {
        {"acid_ant", acid_ant},
        {"bloated_bedbug", bloated_bedbug},
        {"dung_beetle", dung_beetle},
        {"engorged_tick", engorged_tick},
        {"famished_tick", famished_tick},
        {"foraging_maggot", foraging_maggot},
        {"infected_mouse", infected_mouse},
        {"lava_ant", lava_ant},
        {"mantis", mantis},
        {"mawing_beaver", mawing_beaver},
        {"plague_bat", plague_bat},
        {"rhino_beetle", rhino_beetle},
        {"swooping_bat", swooping_bat},
        {"tainted_cockroach", tainted_cockroach},
        {"tunneling_mole", tunneling_mole}
    };

    while (std::getline(std::cin, unit_state_str)) {
        unit_state = nlohmann::json::parse(unit_state_str);
        unit_id = unit_state["self"]["id"];
        std::string unit_type = unit_state["self"]["type"];

        auto it = unit_functions.find(unit_type);
        if (it != unit_functions.end()) {

            UnitFunction unit_function = it->second;
            std::tuple<int, int, float, std::string> instruction = unit_function(unit_state, memory[unit_id]);

            std::cout << unit_id << ":" << "attack" << std::endl;
            std::cout.flush();
        } else {
            std::cerr << "Unknown unit type: " << unit_type << std::endl;
        }
    }

    return 0;
}
